gobblin-api/src/main/java/gobblin/source/workunit/Extract.java

```diff
@@ -81,14 +88,47 @@ public Extract(SourceState state, TableType type, String namespace, String table
       }
 
       // Setting full drop date if not already specified, the value can still be overridden if required.
-      if (state.getPropAsBoolean(ConfigurationKeys.EXTRACT_IS_FULL_KEY) && !state
-          .contains(ConfigurationKeys.EXTRACT_FULL_RUN_TIME_KEY)) {
+      if (state.getPropAsBoolean(ConfigurationKeys.EXTRACT_IS_FULL_KEY)
+          && !state.contains(ConfigurationKeys.EXTRACT_FULL_RUN_TIME_KEY)) {
         super.setProp(ConfigurationKeys.EXTRACT_FULL_RUN_TIME_KEY, System.currentTimeMillis());
       }
     }
   }
 
   /**
+   * Constructor.
+   *
+   * @param type {@link TableType}
+   * @param namespace dot separated namespace path
+   * @param table table name
+   */
+  public Extract(TableType type, String namespace, String table) {
+    this(new SourceState(), type, namespace, table);
+  }
+
+  /**
+   * Factory method. Returns a unique {@link Extract} instance.
+   *
+   * @param type {@link TableType}
+   * @param namespace dot separated namespace path
+   * @param table table name
+   * @return a unique {@link Extract} instance
+   */
+  public static Extract getUniqueInstance(TableType type, String namespace, String table) {
```

Comment:
This looks like a copy/paste of gobblin.configuration.SourceState.createExtract(TableType, String, String) which is not good. We need to remove this function and that static CREATED_INSTANCES object. The extract creation should be owned by the Source.

1) Please move the CREATED_INSTANCES map and this method into a new interface/class ExtractFactory/ExtractFactoryImpl or something along those lines. 

2) Have AbstractSource create/own an instance of ExtractFactoryImpl.

3) Have the KafkaSource use the factory provided by AbstractSource